File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
// 흠;;
// 트럭이 들어온지 몇초 지났는지를 세는 stack2를 하나 더 만들까?
// 만약, 아래에서 stack.push가 되면, stack2도 1을 푸쉬하는 것
// 매.초.마다 stack2는 항상 모든 요소가 늘어가고, 만약 length를 초과하면 stack1을 없애기.

function solution(bridge_length, weight, truck_weights) {
  let stack = []
  let onStack = []
  let cnt = 0
  let stackSum = 0

  while (truck_weights.length || stack.length) {
    cnt ++; // 매초
    
    onStack = onStack.map(item => item+1)
    if (onStack[0] === bridge_length) {
      stackSum -= stack[0]
      stack.shift()
      onStack.shift()
    }
    
    if (truck_weights[0]+stackSum <= weight) {
      stackSum += truck_weights[0]
      stack.push(truck_weights.shift()) // 빼고, 넣기
      onStack.push(0)
    };
  }

  return cnt
}

// console.log(solution(2,10,[7,4,5,6]))
console.log(solution(4,2,[1, 1, 1, 1])) // 10
console.log(solution(5,5,[1,1,1,1,1,2,2])) // 14
console.log(solution(7,7,[1,1,1,1,1,3,3])) // 18
console.log(solution(5,5,[1,1,1,1,1,2,2,2,2])) // 19
console.log(solution(5,5,[2, 2, 2, 2, 1, 1, 1, 1, 1])) // 19

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# bridge_length = 트럭이 다리 건너는 시간인듯
# 1중포문에 while문 조금 곁들여서 가능할 것 같은데 해보기
# 근데, 다리를 건너고 있는 트럭의 '현재 상태'를 어떻게 표현하지?
    # 다리에 올라온(on) 트럭은 매초마다 진행을 해야함
# 아 근데 이러면 'times변수'관리가 좀 힘들거 같은데..

# 그냥 아예 처음부터 while로 시작하고, while의 매 반복은 +1초로 하자
def solution(bridge_length, weight, truck_weights):
    times = 1
    stack, running_stack = [], []

    # while한번에 1초 지남
    while len(truck_weights) != 0 or len(stack) != 0: 
        # step1: stack에 넣기
        if len(truck_weights) != 0 and sum(stack)+truck_weights[0] <= weight:
            running_stack.append(bridge_length) # 아래에서 바로 뺄거니까 +1해서 넣음
            stack.append(truck_weights.pop(0))
        
        # step2: 다리 건너기(진행)
            # 단, 최근에 들어온건 진행 노노 -> 위에서 length+1로 해결함
        running_stack = [running_stack[x]-1 for x in range(len(running_stack))]
        
        # step3: 다 건넌건 빼기
        if running_stack[0] == 0:
            stack.pop(0)
            running_stack.pop(0)
        
        # step4: 시간 세기
        times += 1
            

    return times


print(solution(2,10,[7,4,5,6]))
print(solution(100,100,[10]))
print(solution(100,100,[10,10,10,10,10,10,10,10,10,10]))